### PR TITLE
feat: add lines property in feret diameter

### PR DESCRIPTION
**Pull request by EscapedGibbon**

### CODE REVIEW

File: src/maskAnalysis/__tests__/getFeret.test.ts [hunk 1]
```diff
@@ -47,6 +70,28 @@ test('mask with only 1 pixel', () => {
...
+            { column: 2, row: 0 },
+          ],
+          [
+            { column: 1, row: 1 },
+            { column: 2, row: 1 },
+          ],
+        ],
+        maxDiameter: [
+          [
+            { column: 1.5000000000000002, row: -0.5 },
```
Comment (stropitek) [inline src/maskAnalysis/__tests__/getFeret.test.ts:86]: The principle of using `toBedeepCloseTo` is to not have to specify all the decimals in tests

File: src/maskAnalysis/__tests__/getFeret.test.ts [hunk 2]
```diff
@@ -19,6 +20,28 @@ test('empty mask', () => {
           { column: 0, row: 0 },
         ],
       },
+      lines: {
```
Comment (lpatiny) [inline src/maskAnalysis/__tests__/getFeret.test.ts:23]: lines property should rather be added in the existing minDiameter property

File: src/maskAnalysis/getFeret.ts [hunk 3]
```diff
@@ -126,9 +156,272 @@ export function getFeret(mask: Mask): Feret {
...
+/**
+ * Calculates calliper lines for Feret minDiameter
+ *
+ * @param HullPoints - points that constitute convexHull
+ * @param index - saved indexes of feretDiameter max/min points
+ * @param angle - angle of feretDiameter max/min
+ * @returns array of arrays with two points
+ */
+
+function getMinCalliper(HullPoints: Point[], index: number[], angle: number) {
```
Comment (lpatiny) [inline src/maskAnalysis/getFeret.ts:214]: Should be hullPoints (first letter should be lowercase)

**Review by lpatiny — CHANGES_REQUESTED** [review]
In order to avoid out of bounds of array you can use modulo

```
const points = [10, 20, 30, 40, 50];

for (let i = 0; i < points.length; i++) {
  console.log(
    points[i],
    points[(i + 1) % points.length],
    points[(i + 2) % points.length]
  );
}
```

File: src/maskAnalysis/getFeret.ts [hunk 4]
```diff
@@ -73,7 +99,7 @@ export function getFeret(mask: Mask): Feret {
   let minWidth = Number.POSITIVE_INFINITY;
   let minWidthAngle = 0;
   let minLinePoints: Point[] = [];
```
Comment (lpatiny) [inline src/maskAnalysis/getFeret.ts:101]: We only need during the search for the minFeretDiameter to keep the minX and maxX in order to calculate the minCalliperLines. You can name this variable minCalliperXs.

File: src/maskAnalysis/getFeret.ts [hunk 5]
```diff
@@ -82,23 +108,25 @@ export function getFeret(mask: Mask): Feret {
...
         currentWidth = absWidth;
         currentMinLinePoints = [rotatedPoints[i], rotatedPoints[j]];
+        currentMinLineIndex = [i, j];
       }
     }
     if (currentWidth < minWidth) {
       minWidth = currentWidth;
       minWidthAngle = angle;
       minLinePoints = currentMinLinePoints;
+      minLineIndex = currentMinLineIndex;
```
Comment (lpatiny) [inline src/maskAnalysis/getFeret.ts:126]: You should update here the minCalliperXs (in reality minX and maxX)

File: src/maskAnalysis/getFeret.ts [hunk 6]
```diff
@@ -82,23 +108,25 @@ export function getFeret(mask: Mask): Feret {
...
     }
   }
+
   const minDiameter = {
     points: rotate(minWidthAngle, minLinePoints),
     length: minWidth,
```
Comment (lpatiny) [inline src/maskAnalysis/getFeret.ts:132]: Based on minCalliperXs you can directly calculate the 2 calliper lines.

**Review by lpatiny — CHANGES_REQUESTED** [review]
In order to avoid out of bounds of array you can use modulo

```
const points = [10, 20, 30, 40, 50];

for (let i = 0; i < points.length; i++) {
  console.log(
    points[i],
    points[(i + 1) % points.length],
    points[(i + 2) % points.length]
  );
}
```

File: src/maskAnalysis/getFeret.ts [hunk 7]
```diff
@@ -97,17 +129,29 @@ export function getFeret(mask: Mask): Feret {
       minWidth = currentWidth;
       minWidthAngle = angle;
       minLinePoints = currentMinLinePoints;
+      const currentMin = findPointsOfExtremeColumns(rotatedPoints).minIndex;
+      const currentMax = findPointsOfExtremeColumns(rotatedPoints).maxIndex;
```
Comment (lpatiny) [inline src/maskAnalysis/getFeret.ts:133]: const {minIndex: currentMin, maxIndex: currentMax} = find...

File: src/maskAnalysis/getFeret.ts [hunk 8]
```diff
@@ -132,3 +189,94 @@ export function getFeret(mask: Mask): Feret {
...
   };
 }
+
+function findPointsOfExtremeColumns(points: Point[]): {
+  minIndex: number;
+  maxIndex: number;
+} {
+  let max = 0;
+  let min = 0;
+  for (let i = 0; i < points.length; i++) {
```
Comment (lpatiny) [inline src/maskAnalysis/getFeret.ts:199]: Why 2 loops ? One should be enough

File: src/maskAnalysis/getFeret.ts [hunk 9]
```diff
@@ -132,3 +189,94 @@ export function getFeret(mask: Mask): Feret {
...
+  }
+  return { minIndex: min, maxIndex: max };
+}
+function findPointsOfExtremeRows(points: Point[]): {
+  minIndex: number;
+  maxIndex: number;
+} {
+  let max = 0;
+  let min = 0;
+  for (let i = 0; i < points.length; i++) {
```
Comment (lpatiny) [inline src/maskAnalysis/getFeret.ts:217]: One loop should be enough

File: src/maskAnalysis/__tests__/getFeret.test.ts [hunk 10]
```diff
@@ -27,8 +48,8 @@ test('empty mask', () => {
 test('mask with only 1 pixel', () => {
   let mask = testUtils.createMask(`0 1 0`);
   const result = mask.getFeret();
-  // the minimum diameter points are not as expected,
-  // but this is a very edge case so it does not matter too much
+  //   // the minimum diameter points are not as expected,
+  //   // but this is a very edge case so it does not matter too much
```
Comment (stropitek) [inline src/maskAnalysis/__tests__/getFeret.test.ts:52]: ?

File: src/maskAnalysis/__tests__/getFeret.test.ts [hunk 11]
```diff
@@ -62,21 +103,42 @@ test('mask 3x3', () => {
...
+        lines: [
+          [
+            { column: -0.5000000000000001, row: 1.4999999999999998 },
+            { column: 1.5, row: 3.5 },
+          ],
+          [
+            { column: 1.5, row: -0.5000000000000002 },
+            { column: 3.5, row: 1.5000000000000002 },
+          ],
+        ],
```
Comment (stropitek) [inline src/maskAnalysis/__tests__/getFeret.test.ts:121]: Here as well, you don't have to write out every decimal

File: src/maskAnalysis/__tests__/getFeret.test.ts [hunk 12]
```diff
@@ -62,21 +103,42 @@ test('mask 3x3', () => {
...
+        lines: [
+          [
+            { column: 0.5000000000000002, row: -0.5000000000000001 },
+            { column: -0.5, row: 2.4999999999999996 },
+          ],
+          [
+            { column: 3.5, row: 0.5000000000000001 },
+            { column: 2.4999999999999996, row: 3.5 },
+          ],
+        ],
```
Comment (stropitek) [inline src/maskAnalysis/__tests__/getFeret.test.ts:139]: ditto

File: src/maskAnalysis/getFeret.ts [hunk 13]
```diff
@@ -97,17 +128,29 @@ export function getFeret(mask: Mask): Feret {
       minWidth = currentWidth;
       minWidthAngle = angle;
       minLinePoints = currentMinLinePoints;
+      const { minIndex: currentMin, maxIndex: currentMax } =
+        findPointsOfExtremeColumns(rotatedPoints);
+      minLines = formMinLines(
```
Comment (lpatiny) [inline src/maskAnalysis/getFeret.ts:133]: Please rename to `getMinLines`

File: src/maskAnalysis/getFeret.ts [hunk 14]
```diff
@@ -73,7 +96,16 @@ export function getFeret(mask: Mask): Feret {
   let minWidth = Number.POSITIVE_INFINITY;
   let minWidthAngle = 0;
   let minLinePoints: Point[] = [];
-
+  let minLines: [[Point, Point], [Point, Point]] = [
```
Comment (lpatiny) [inline src/maskAnalysis/getFeret.ts:99]: You should not define minLines if you are regenerating a new object later in your function.

File: src/maskAnalysis/getFeret.ts [hunk 15]
```diff
@@ -132,3 +188,95 @@ export function getFeret(mask: Mask): Feret {
     aspectRatio: minDiameter.length / maxDiameter.length,
   };
 }
+
+function findPointsOfExtremeColumns(points: Point[]): {
+  minIndex: number;
+  maxIndex: number;
+} {
+  let max = 0;
```
Comment (lpatiny) [inline src/maskAnalysis/getFeret.ts:196]: The variables should be named minIndex and maxIndex.

File: src/maskAnalysis/getFeret.ts [hunk 16]
```diff
@@ -132,3 +188,95 @@ export function getFeret(mask: Mask): Feret {
...
+      min = i;
+    }
+  }
+  return { minIndex: min, maxIndex: max };
+}
+function findPointsOfExtremeRows(points: Point[]): {
+  minIndex: number;
+  maxIndex: number;
+} {
+  let max = 0;
```
Comment (lpatiny) [inline src/maskAnalysis/getFeret.ts:213]: Also minIndex and maxIndex

File: src/maskAnalysis/getFeret.ts [hunk 17]
```diff
@@ -132,3 +188,95 @@ export function getFeret(mask: Mask): Feret {
...
+      row: feretPoints[1].row,
+    },
+  ];
+
+  return [rotate(angle, minLine1), rotate(angle, minLine2)] as [
+    [Point, Point],
+    [Point, Point],
+  ];
+}
+function formMaxLines(
```
Comment (lpatiny) [inline src/maskAnalysis/getFeret.ts:256]: Please also rename this to `getMaxLines`

**Review by lpatiny — CHANGES_REQUESTED** [review]
Still some changes to do

File: src/maskAnalysis/__tests__/getFeret.test.ts (hunk 10 above)
Comment (EscapedGibbon) [reply src/maskAnalysis/__tests__/getFeret.test.ts:52]: It was there as a test. I tailored the values so that the test passes, since I assumed that it's not that important. Still i believe the test will provide a plausible result. Should I remove the comments?
